### PR TITLE
Allow testing of pre-release agent zip in hosted pools

### PR DESCRIPTION
**Pull request by johnterickson**
For quick iteration of performance testing, we need to run custom bits in the actual production environment.  This PR allows a contributor to store their agent.zip in a publicly accessible HTTP endpoint and the agent will update to it.

### CODE REVIEW

File: src/Agent.Listener/SelfUpdater.cs [hunk 1]
```diff
@@ -36,14 +36,42 @@ public override void Initialize(IHostContext hostContext)
...
-            var settings = configStore.GetSettings();
-            _poolId = settings.PoolId;
-            _agentId = settings.AgentId;
+            _agentSettings = configStore.GetSettings();
         }
 
         public async Task<bool> SelfUpdate(AgentRefreshMessage updateMessage, IJobDispatcher jobDispatcher, bool restartInteractiveAgent, CancellationToken token)
         {
-            if (!await UpdateNeeded(updateMessage.TargetVersion, token))
+            var extensionManager = HostContext.GetService<IExtensionManager>();
```
Comment (johnterickson) [inline src/Agent.Listener/SelfUpdater.cs]: @TingluoHuang  Does this get called normally? Will it at least get called if I have a task that requires a higher versioned agent than is currently running?

**Comment by johnterickson** [comment]
FYI @willsmythe @owenhuynMSFT @fadnavistanmay @MihaifGit

**Comment by johnterickson** [comment]
@willsmythe pointed out we might be able to be more precise by hooking the load of the plugin 🤔https://github.com/microsoft/azure-pipelines-agent/blob/688423668d50a7ce7d047945419127cbf3187cde/src/Agent.Worker/AgentPluginManager.cs#L138-L139

**Comment by TingluoHuang** [comment]
@johnterickson why are we doing this? you can test the agent in appfarbric with MMS

**Pull request closed without merging**